From 1e01ffb96cc927a202af6eda30c57cd8dc70398b Mon Sep 17 00:00:00 2001
From: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Date: Fri, 1 Jun 2012 17:10:11 +0200
Subject: [PATCH 06/78] wifi: Send a scan complete to mac layer when cancel
 scan in off state

BZ: 34845

In some rare occasions, it is possible to have a wlan firmware recovery
while a hw scan is in progress. When this occurs, the driver never tells
the mac layer that it should stop the scan operation, leaving the mac layer
and the driver in opposite scanning states.
On following hw restart, the mac layer is rejecting all scan operations
because it is in scanning state already.
Fix this by complying to mac layer requirement, ie call ieee80211_scan_completed
in op_hw_cancel_scan even when wl12xx state if OFF (in the middle of a recovery)

Change-Id: I8ee13f81bcad232bf4cd4a76895fe2c4285637e6
Signed-off-by: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Reviewed-on: http://android.intel.com:8080/50968
Reviewed-by: Lemetayer, Jean-marieX <jean-mariex.lemetayer@intel.com>
Reviewed-by: Bachot, Jean-michel <jean-michel.bachot@intel.com>
Reviewed-by: Trivelly, Jean <jean.trivelly@intel.com>
Reviewed-by: Tricot, HonoreX <honorex.tricot@intel.com>
Reviewed-by: Regairaz, Julie <julie.regairaz@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/59012
Reviewed-on: http://android.intel.com:8080/77963
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c |   10 ++++++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index 3434ac0..096389f 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -3601,9 +3601,15 @@ static void wl1271_op_cancel_hw_scan(struct ieee80211_hw *hw,
 
 	mutex_lock(&wl->mutex);
 
-	if (wl->state == WL1271_STATE_OFF)
+	if (wl->state == WL1271_STATE_OFF) {
+		/*
+		If we get here, it means we have a FW recovery with a pending scan
+		Alert the mac layer that scan is complete, as required in
+		ieee80211_scan_cancel()
+		*/
+		ieee80211_scan_completed(wl->hw, true);
 		goto out;
-
+	}
 	if (wl->scan.state == WL1271_SCAN_STATE_IDLE)
 		goto out;
 
-- 
1.7.9

